MTV-2446 | OCP to OCP migration fail due to MAC address collisions

Hazanel opened the pull request:
Issue:
OCP-to-OCP migrations failed with: "admission webhook 'mutatevirtualmachines.kubemacpool.io' denied the request: Failed to allocate mac to the vm object: failed to allocate requested mac address"

Fix:
Adding automatic kubemacpool namespace exclusion logic to both cold and live migration paths. Includes automatic cleanup that removes the exclusion label after migration completion, cancellation, or archival to restore the namespace to its original state.

Ref: https://issues.redhat.com/browse/MTV-2446

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Per-namespace kubemacpool exclusion for same-cluster OpenShift migrations; auto-applied before VM creation and removed on archive/cancel/finish.

* **Improvements**
  * Manifest processing accepts multiple formats and returns the first VM found.
  * VM creation now preserves/merges annotations and derives run state from target/restore power settings.
  * KubeVirt types auto-registered for broader compatibility.

* **Observability**
  * Enhanced logging for manifest decoding, retrieval, and namespace-exclusion flows.

* **Tests**
  * Extensive tests covering manifest extraction (HTTP/TLS/token), namespace exclusion, and lifecycle cleanup.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mansam reviewed (changes requested):
I'm still going through these changes in detail, but it doesn't look like these changes belong in the plancontext.

Hazanel commented:
@mansam  - I've added a new file.
Basically it's a common code and I was trying to avoid duplication.
I hope that's what you were referring to

mansam commented on pkg/controller/plan/adapter/ocp/builder_test.go:
```diff
@@ -0,0 +1,367 @@
+	}
+
+	// Simulate the builder's DeepCopy approach
+	targetVmSpec := sourceVm.Spec.DeepCopy()
```
What is the point of these deepcopy tests? Contrary to what the comment says, these tests aren't testing anything the builder does; they're just testing that the DeepCopy method (which isn't MTV code) works.

mansam reviewed (changes requested):
As I understand it, this is silently disabling automatic MAC address assignment via kubemacpool for any namespace where we are migrating OCP-to-OCP. I do not believe that this is an appropriate solution to the problem of MAC conflicts during migration as-written.

1. The namespace's original configuration needs to be restored post-migration. We don't know how the namespace was being used originally or how it's going to be used going forward so it's important that we set the configuration back to how it originally was after migration so that we don't break any user-requested VM deployments. We need to keep track of the original annotation if there already was one, and need to keep track of plans migrating to that namespace via some kind of annotation and only reset it once the last plan is done with it, using some mechanism (conflict/retry) to avoid race conditions.
2. We should be proactively looking for these conflicts and communicating them to the user. MAC conflicts like this need to be surfaced in the form of validations or concerns, so that the end user is aware of the problem and can take appropriate action or at least not be surprised when MTV alters the namespace configuration. We need to make sure the user knows that creating VMs in the destination namespace while MTV is migrating to it is most likely going to result in undesired behavior.
3. This doesn't belong in the ensurer/builder. Configuring the destination namespace with this annotation (and then undoing it) should probably be done as part of setup and teardown in the migration process rather than silently as a side effect while creating a VM resource.
4. I think we only want this to happen for namespace-to-namespace migrations. A MAC address conflict while migrating to another cluster would indicate a real problem to me and we shouldn't try to paper over it.
5. We might want to consider having an option to block a migration over a MAC conflict rather than alter the namespace.

Nitpicks:

1. We don't need checks that the destination provider is OCP. The destination is always necessarily an OCP provider.
2. Please try to keep unrelated changes in separate PRs or at least in separate commits. The changes to the OCP builder don't seem to have any relationship to MAC address conflict issue.
3. There's no reason to add unit tests that only exercise library code. Checking that things are copied by DeepCopy is unnecessary; we did not write DeepCopy. The comments suggest this was done to "simulate" builder behavior, but unit testing a "simulation" of code is not useful since there's no relationship between the code being tested and the actual implementation.

Hazanel committed:
Resolves: MTV-2446 | OCP to OCP migration fail due to MAC address collisions

Issue:
OCP-to-OCP migrations failed with: "admission webhook 'mutatevirtualmachines.kubemacpool.io' denied the request: Failed to allocate mac to the vm object: failed to allocate requested mac address"

Fix:
Added ensureNamespaceKubemacpoolExclusion() in kubevirt.EnsureVM()
Updated both regular and live migration code paths (kubevirt + ensurer)
Added comprehensive unit tests for namespace exclusion logic

Ref: https://issues.redhat.com/browse/MTV-2446

Signed-off-by: Elad Hazan <ehazan@redhat.com>

Hazanel commented:
1. Splitted into 2 logical commit 

2. You’re correct that this disables kubemacpool’s automatic MAC assignment for the target namespace — but only temporarily and only for the duration of an OCP-to-OCP migration within the same cluster.

3. The intent here is to prevent MAC conflicts that occur when restoring VMs with pre-existing MACs into a namespace where kubemacpool would otherwise reassign them. Without this step, kubemacpool may generate duplicate MACs, resulting in networking issues for migrated workloads.

4. For cross-cluster migrations, we intentionally do not suppress kubemacpool’s allocation and instead surface a warning about possible MAC conflicts, so administrators can address them explicitly.
 
5. This label change is never permanent — as part of the migration completion flow, we remove mutatevirtualmachines.kubemacpool.io=ignore and restore kubemacpool’s default behavior for that namespace.
 https://docs.redhat.com/en/documentation/openshift_container_platform/4.8/html-single/openshift_virtualization/index#virt-4-8-changes

ekinmeroglu commented:
One suggestion - if I got it right, this code only allows NS to NS migrations when both the source and target are `host` - i.e. the local cluster.  I think remote OCP providers should also be covered and fine, as long as source and target is the same OCP cluster. Having a "head" cluster hosting forklift/MTV and using this single instance to execute migrations in multiple separate clusters is a very common usage pattern in large installations.

And a very big +1 to @mansam 's comment earlier (The namespace's **original** configuration needs to be restored post-migration). The user might already have this label on the target namespace, so we shouldn't blindly remove the label after a successful migration.  Likewise, having this as a configurable option would be great.

Hazanel commented on pkg/controller/plan/namespace/kubemacpool.go:
```diff
@@ -0,0 +1,322 @@
+// KubemacpoolIgnoreLabelValue is the value that tells kubemacpool to ignore a namespace.
+const KubemacpoolIgnoreLabelValue = "ignore"
+
+// IsSameClusterMigration determines if source and destination providers point to the same OCP cluster.
+// This covers both local cluster migrations (both IsHost()) and remote cluster migrations (same URL).
+func IsSameClusterMigration(sourceProvider, destProvider *v1beta1.Provider) bool {
+	// Both must be OpenShift providers
+	if sourceProvider.Type() != v1beta1.OpenShift || destProvider.Type() != v1beta1.OpenShift {
+		return false
+	}
+
+	// Case 1: Both are host providers (local cluster)
+	if sourceProvider.IsHost() && destProvider.IsHost() {
+		return true
+	}
+
+	// Case 2: Both point to the same remote cluster (same URL)
+	if sourceProvider.Spec.URL != "" && destProvider.Spec.URL != "" {
+		return sourceProvider.Spec.URL == destProvider.Spec.URL
+	}
+
+	// Mixed case: one is host, one is remote - this is cross-cluster
+	return false
+}
```
you didn't draft a tiny normalizer (parse + strings.TrimSuffix + default port elision)

Hazanel commented on pkg/controller/plan/kubevirt.go:
```diff
@@ -1537,11 +1551,32 @@ func (r *KubeVirt) virtualMachine(vm *plan.VMStatus, sortVolumesByLibvirt bool)
-		object.ObjectMeta.Annotations = annotations
+	}
+
+	// MAC address conflicts are handled by namespace-level kubemacpool exclusion in EnsureVM()
+	// This follows Red Hat OpenShift Virtualization best practices for production environments
+	// Reference: https://docs.redhat.com/en/documentation/openshift_container_platform/4.8/html-single/openshift_virtualization/index#virt-4-8-changes
+
```
Every link will have an exact version - skipping this comment

mnecas commented:
closing as stale
